pkcs11/src/new/mod.rs

```diff
@@ -23,6 +31,10 @@ pub struct Pkcs11 {
     // valid.
     _pkcs11_lib: pkcs11_sys::Pkcs11,
     function_list: pkcs11_sys::_CK_FUNCTION_LIST,
+    // Handle of sessions currently logged in per slot. This is used for logging in and out.
+    logged_sessions: Mutex<HashMap<Slot, HashSet<CK_SESSION_HANDLE>>>,
+    // Pin per slot, will be used for login. Ideally this should also be filtered by user type.
+    pins: RwLock<HashMap<Slot, SecretVec<u8>>>,
```

Comment:
Now that I see it like this, I could group the two:
```rust
Mutex<HashMap<Slot, (SecretVec<u8>, HashSet<CK_SESSION_HANDLE>)>>
```

Reply:
but not sure if it's good

Reply:
What about 
```rust
HashMap<Slot, (SecretVec<u8>, Mutex<HashSet<CK_SESSION_HANDLE>>)>
```

Reply:
We need a `Mutex` around everything unfortunately to prevent having any function with a mutable reference to `self`, otherwise it's a pain when using the spec as you can't do everything in the same lifetime 😢

Reply:
tbf I am happy with the current one, we can always optimise it later or when we integrate in PArsec

Reply:
> We need a `Mutex` around everything unfortunately to prevent having any function with a mutable reference to `self`

Do you mean methods for setting the PIN? I'm not sure if blocking across all `Slot` instances is needed (as in, per the spec), wouldn't want to make that much slower, but if there's no good solution I'm ok with that.

Reply:
> Do you mean methods for setting the PIN?

Yes.

>  I'm not sure if blocking across all Slot instances is needed (as in, per the spec), wouldn't want to make that much slower, but if there's no good solution I'm ok with that.

Yes I see what you mean. The problem is that you need to modify the `HashMap` in order to add the `Slot` key in the first place... Unless you do everything in the constructor and build the HashMap then will all current possible slots. Then I agree you can only put the `Mutex` on the values. Maybe an optimization to see later.

Reply:
Hmm, yes, I see... I think you'd have another problem if you bundled them like that - you then can't remove the PIN too easily (or do you just clear the `Secret` storing it?). Because if you call `clear_pin` you can't just delete the whole entry, you need to keep the logged in session set.